src/machine/vm.cr

```diff
@@ -1082,6 +1082,11 @@ module VM
         reg = Register.new stack_pop UInt8
         value = reg_read Int32, reg
         STDOUT.puts "#{value}"
+      when Syscall::READ
+        char = STDIN.raw &.read_char
+        if char.is_a?(Char)
+          reg_write Register::R0.byte, char.ord
```

Comment:
[`char.ord`](https://crystal-lang.org/api/master/Char.html#ord%3AInt32-instance-method) returns an `Int32`, so you'd have to write this into the `Register::R0.dword` register. Besides that, maybe the user should be able to provide a custom register where he wants the result to end up in?

I was thinking of something along the lines of the `sys_puts` syscall.

```crystal
when Syscall::PUTS
  reg = Register.new stack_pop UInt8 # A register identifier is popped off the stack here
  value = reg_read Int32, reg
  STDOUT.puts "#{value}"
```

You could then change your `sys_read` implementation to:

```crystal
when Syscall::READ
  reg = Register.new stack_pop UInt8
  char = STDIN.raw &.read_char
  if char.is_a?(Char)
    reg_write reg, char.ord
  end
```